### PR TITLE
Layers UI Rework

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [x] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
This change is based off the design discussion in #1127 
It implements new UI components to bring up the consistency to that of the Data access view. 
image

Error handling changes have been implemented within each query dialog based on recent changes to data access view error handling. The image below shows an incorrectly parsed query int he transaction section. When this occurs, the layer is disabled from being toggled, and the titlepane goes red as well as the specific query entry field. 
image

Outstanding work checklist:
* [x] Refresh layers without recreation of each titlepane object
* [x] Ensure layer 0 is toggled when no other layer is enabled (visible feedback to the user that the complete graph is shown)
* [x] Disable the checkbox in layer 0
* [x] Correctly refresh the layers view when the current graph is switched or closed.

Extended:
* [ ] LayerAttributes which track graph changes and update the graph if necessary are not getting stored.
* [x] Toggling a layer more than once doesn't show correct results due to the modcount not changing (code not thinking it needs to recalculate)

### Alternate Designs
Alternate designs have been discussed in detail within the main issue ticket. Implementation alternatives have not been discussed as yet.

### Why Should This Be In Core?
Enhances consistency of the UI elements throughout different Constellation Views
Enhances readability of queries and layers view workflow

### Benefits
as above

### Possible Drawbacks
none realised

### Verification Process
WIP

### Applicable Issues
#1127

### CODE REVIEW

**Comment by aldebaran30701** [comment]
I haven't managed to resolve all bugs, but the checkboxes in the description show the remaining tasks. 

To add onto this, the recent parameter values have not been fully implemented within the QueryInputPane. 

I also noticed there was one case when I entered an invalid query into Vertex section and both vertex and transaction query entries went red. Only the vertex field should do so.

File: CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/utilities/ExpressionUtilities.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ *
+ * A collection of utility methods to assist in using the Expressions framework
+ *
+ * @author aldebaran30701
+ */
+public final class ExpressionUtilities {
+
+    private ExpressionUtilities() {
+        // intentionally left blank
```
Comment (antares1470) [inline CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/utilities/ExpressionUtilities.java:29]: Are you able to throw an `IllegalStateException` here instead (similar to other utilities classes)

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 2]
```diff
@@ -292,18 +293,123 @@ public void shuffleElementBitmasks(final int currentIndex) {
...
+
+        if (layerCount <= BitMaskQueryCollection.MAX_QUERY_AMT) {
+            Query vxQuery = new Query(GraphElementType.VERTEX, "");
+            getVxQueryCollection().add(vxQuery, layerCount + 1, null);
+            Query txQuery = new Query(GraphElementType.TRANSACTION, "");
+            getTxQueryCollection().add(txQuery, layerCount + 1, null);
+            writeState();
+        } else {
+            NotifyDisplayer.display("You cannot have more than " + BitMaskQueryCollection.MAX_QUERY_AMT + " layers", NotifyDescriptor.WARNING_MESSAGE);
+            LOGGER.log(Level.INFO, "Layer count maximum reached. Maximum is currently: {0}", BitMaskQueryCollection.MAX_QUERY_AMT);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:316]: Are you able to change the level to warning (just based on the line above it)

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 3]
```diff
@@ -292,18 +293,123 @@ public void shuffleElementBitmasks(final int currentIndex) {
...
+    /**
+     * Create a layer by writing data to the UI and saving that state
+     */
+    public void createLayer() {
+        final int layerCount = Math.max(getTxQueryCollection().getHighestQueryIndex(), getVxQueryCollection().getHighestQueryIndex());
+
+        if (layerCount <= BitMaskQueryCollection.MAX_QUERY_AMT) {
+            Query vxQuery = new Query(GraphElementType.VERTEX, "");
+            getVxQueryCollection().add(vxQuery, layerCount + 1, null);
+            Query txQuery = new Query(GraphElementType.TRANSACTION, "");
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:311]: both queries can be final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 4]
```diff
@@ -359,17 +465,18 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
...
                 currentState = new LayersViewState();
-                currentState.setVxLayers(BitMaskQueryCollection.DEFAULT_VX_QUERIES);
-                currentState.setTxLayers(BitMaskQueryCollection.DEFAULT_TX_QUERIES);
+                currentState.setVxLayers(BitMaskQueryCollection.getDefaultVxQueries());
+                currentState.setTxLayers(BitMaskQueryCollection.getDefaultTxQueries());
             } else {
                 currentState = new LayersViewState(currentState);
             }
 
+            LOGGER.log(Level.SEVERE, "is query visible already?: " + currentState.getVxQueriesCollection().getQuery(1).isVisible());
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:474]: omit this log

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    
+    private final int layerId;
+    private boolean isValid;
+
+    public LayerTitlePane(final int layerId, final String layerName, final BitMaskQuery query) {
+        this.query = query;
+        isValid = testQueryValidity(query.getQueryString());
+        enabled = new CheckBox();
+        
+        enabledChanged = (ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) -> {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java:77]: either make the lambda variables final or omit the explicit types

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        return isValid;
+    }
+    
+    public void setValidity(final boolean isValid) {
+        this.isValid = isValid;
+        recolourLayer();
+    }
+    
+    public void recheckValidity() {
+        boolean validity = query.getQueryString() == null || ExpressionUtilities.testQueryValidity(query.getQueryString());
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java:255]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        this.isValid = validity;
+        setValidity(validity);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java:257]: given `setValidity(validity)` will do exactly what the line above it is doing, I think that line above can be removed

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+    /**
+     * Refresh the layers pane and update each layer
+     *
+     * @param vxLayers
+     * @param txLayers
+     */
+    public synchronized void setLayers(final BitMaskQuery[] vxLayers, final BitMaskQuery[] txLayers) {
+        CountDownLatch cdl1 = new CountDownLatch(1);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:133]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+            createLayers(oldLayers, vxLayers, txLayers);
+            attributeScrollPane.setContent(oldLayers);
+            this.layersViewVBox = new VBox(layersHeading, attributeScrollPane, options);
+            // trigger refresh using enabled method
+            setEnabled(true);
+            cdl1.countDown();
+        });
+        try {
+            cdl1.await();
+        } catch (InterruptedException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:151]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            LOGGER.log(Level.WARNING, ex.getMessage());
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:152]: can you change this level to severe?

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * Set the layers to the defaults.
+     */
+    public synchronized void setDefaultLayers() {
+        CountDownLatch cdl1 = new CountDownLatch(1);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:161]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+            createLayers(oldLayers, controller.getVxQueryCollection().getQueries(), controller.getTxQueryCollection().getQueries());
+            attributeScrollPane.setContent(oldLayers);
+            this.layersViewVBox = new VBox(layersHeading, attributeScrollPane, options);
+            // trigger refresh using enabled method
+            setEnabled(true);
+            cdl1.countDown();
+        });
+        try {
+            cdl1.await();
+        } catch (InterruptedException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:177]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+            LOGGER.log(Level.WARNING, ex.getMessage());
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:178]: severe

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+            // Set query description to either the vertex of transaction description if not null
+            String queryDescription = StringUtils.EMPTY;
+            if (vxQuery != null) {
+                queryDescription = vxQuery.getDescription();
+            } else if (txQuery != null) {
+                queryDescription = txQuery.getDescription();
+            }
+            
+            final Query q;
+            if (StringUtils.isEmpty(txqueryString) && StringUtils.isNotEmpty(vxqueryString)) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:226]: are blank strings meant pass here? Wondering if it should be `isBlank` and `isNotBlank` respectively instead

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+
+            final BitMaskQuery bmq = new BitMaskQuery(q, position, queryDescription);
+            bmq.setVisibility(queryVisibility);
+            
+            if (position == 0 && !isQueryActive) {
+                bmq.setVisibility(!isQueryActive);
+            }
+
+            if (layersnew.getChildren().size() - 1 < queryIndex) {
+                LayerTitlePane tp = new LayerTitlePane(queryIndex,queryDescription, bmq);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:248]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+                layersnew.getChildren().add(queryIndex, tp);
+            } else {
+                LayerTitlePane oldTp = (LayerTitlePane) layersnew.getChildren().remove(queryIndex); // no 0 in list of 0
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:251]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+                oldTp.setDescription(queryDescription);
+                oldTp.setQuery(bmq);
+                layersnew.getChildren().add(queryIndex, oldTp);
+            }
+        }
+
+        // remove the remaining old layers.
+        final int totalLayers = layersnew.getChildren().size();
+        if(totalLayers > iteratorEnd) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:260]: are you able to add a space between `if` and `(` here?

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+            layersnew.getChildren().remove(iteratorEnd, totalLayers);
+        }
+        
+        // set layer 0 selected when no other layer is enabled.
+        LayerTitlePane oldTp = (LayerTitlePane) layersnew.getChildren().remove(0);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayersViewPane.java:265]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/NoGraphPane.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * Create a blank view when the layers view is open without a graph being open at the same time. 
+ * Stops the user from being able to interact with layers view without a graph being open
+ *
+ * @author aldebaran30701
+ */
+public class NoGraphPane extends VBox {
+
+    final Label noGraphLabel;
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/NoGraphPane.java:32]: add an access modifier here (`protected` or `private`, whichever one is more suitable)

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+    private final String parameterId;
+    private final LayerTitlePane parent;
+    private final boolean validityCheckRequired;
+    private List<String> recentValues = new ArrayList<>();
+
+    public QueryInputPane(final LayerTitlePane parent, final String parameter, final String description, final String value, final boolean requiresValidityCheck) {
+        this(parent, parameter, description, value, DEFAULT_WIDTH, null, requiresValidityCheck);
+    }
+
+    public QueryInputPane(final LayerTitlePane parent, final String parameter, final String description, final String value, int defaultWidth, final boolean requiresValidityCheck) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:74]: `defaultWidth` can be final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    /**
+     * Primary constructor
+     *
+     * @param title parameter to link to value
+     * @param defaultWidth default width (in pixels)
+     * @param suggestedHeight suggested hight (in lines)
+     */
+    public QueryInputPane(final LayerTitlePane parent, final String title, final String description, final String value, int defaultWidth, Integer suggestedHeight, final boolean requiresValidityCheck) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:85]: same here

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 22]
```diff
@@ -0,0 +1,316 @@
...
+        recentValuesCombo.setTooltip(new Tooltip("Recent values"));
+        recentValuesCombo.setMaxWidth(5);
+        if (RecentParameterValues.getRecentValues(parameterId) != null) {
+            recentValues = RecentParameterValues.getRecentValues(parameterId);
+        }
+        setRecentValuesCombo(recentValues);
+
+        final ListCell<String> button = new ListCell<String>() {
+            @Override
+            protected void updateItem(String item, boolean empty) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:122]: parameters final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 23]
```diff
@@ -0,0 +1,316 @@
...
+        field.setMinHeight(suggestedHeight);
+        
+
+        field.setPromptText(title);
+        field.setText(value);
+
+        field.setMinWidth(defaultWidth);
+
+        recentValueSelectionListener = (ov, t, t1) -> {
+            String recentValue = recentValuesCombo.getValue();
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:139]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 24]
```diff
@@ -0,0 +1,316 @@
...
+        recentValuesCombo.getSelectionModel().selectedIndexProperty().addListener(recentValueSelectionListener);
+
+        // If parameter is enabled, ensure widget is both enabled and editable.
+        field.setEditable(true);
+        field.setDisable(false);
+        field.setVisible(true);
+        this.setVisible(true);
+        recentValuesCombo.setDisable(false);
+
+        field.addEventFilter(KeyEvent.KEY_PRESSED, (KeyEvent event) -> {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:156]: lambda parameter either final or explicit type removed

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 25]
```diff
@@ -0,0 +1,316 @@
...
+            if (event.getCode() == KeyCode.DELETE) {
+                IndexRange selection = field.getSelection();
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:158]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+            } else if (event.getCode() == KeyCode.ESCAPE) {
+                event.consume();
+            } else {
+                // Do nothing
+            }
+        });
+
+        final Tooltip tooltip = new Tooltip("");
+        tooltip.setStyle("-fx-text-fill: white;");
+        field.focusedProperty().addListener((ObservableValue<? extends Boolean> ov, Boolean t, Boolean t1) -> {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:195]: lambda parameters

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+    /**
+     * Update the field UI to match whether the query is valid
+     * @param isValid
+     */
+    public void setValidity(final boolean isValid) {
+        if (isValid) {
+            field.setId(StringUtils.EMPTY);
+        } else {
+            field.setId(INVALID_ID);
+        }
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/QueryInputPane.java:259]: can you make this a ternary statement?

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/Query.java [hunk 28]
```diff
@@ -59,42 +53,15 @@ public GraphElementType getElementType() {
...
         final Object result = ExpressionCompiler.compileSequenceExpression(expression, variableProvider, index, Operators.getDefault());
 
-        graphId = graph.getId();
-        globalModificationCounter = graph.getGlobalModificationCounter();
-        structureModificationCounter = graph.getStructureModificationCounter();
-        attributeModificationCounter = graph.getAttributeModificationCounter();
-
         attributeIds = variableProvider.getAttributeIds();
-        valueModificationCounters = new long[attributeIds.length];
+        long[] valueModificationCounters = new long[attributeIds.length];
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/Query.java:64]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]
Are you able to add a whatsnew entry for this update. Also thinking it might be good to move this to the main views menu rather than leaving it in experimental if you're able to make that change as well

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java [hunk 29]
```diff
@@ -252,9 +252,8 @@ public void setValidity(final boolean isValid) {
     }
     
     public void recheckValidity() {
-        boolean validity = query.getQueryString() == null || ExpressionUtilities.testQueryValidity(query.getQueryString());
+        final boolean validity = query.getQueryString() == null || ExpressionUtilities.testQueryValidity(query.getQueryString());
         this.isValid = validity;
-        setValidity(validity);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java]: Did you need to the layer re-colour to occur? If so, you probably want to keep the function call and remove the `this.isValid...` line instead

**Review by antares1470 — CHANGES_REQUESTED** [review]
Just did some testing as well. Found a few things:
1) Are you able to update the first screenshot in the help page? This is a screenshot of the old interface.
2) Looks like the right-click context menu may have the menus a little off with an extra layers menu. Are you able to fix that?
LayersContextMenu
3) Looks like graph elements which are added to a layer via query aren't removed when the query is removed (only when the query is changed). Not sure what the intended behaviour is here but I would think those graph elements should be removed from the layer
layersViewQueryIssue

**Comment by Delphinus8821** [comment]
> Just did some testing as well. Found a few things:
> 
> 1. Are you able to update the first screenshot in the help page? This is a screenshot of the old interface.
> 2. Looks like the right-click context menu may have the menus a little off with an extra layers menu. Are you able to fix that?
> 3. Looks like graph elements which are added to a layer via query aren't removed when the query is removed (only when the query is changed). Not sure what the intended behaviour is here but I would think those graph elements should be removed from the layer
          
@antares1470 Points 1 and 3 should now be fixed.
Point 2 is something that has been occurring in the main branch since release v2.5. I will keep looking but at this stage I haven't found a fix for it.

**Comment by antares1470** [comment]
> > Just did some testing as well. Found a few things:
> > 
> > 1. Are you able to update the first screenshot in the help page? This is a screenshot of the old interface.
> > 2. Looks like the right-click context menu may have the menus a little off with an extra layers menu. Are you able to fix that?
> > 3. Looks like graph elements which are added to a layer via query aren't removed when the query is removed (only when the query is changed). Not sure what the intended behaviour is here but I would think those graph elements should be removed from the layer
> 
> @antares1470 Points 1 and 3 should now be fixed. Point 2 is something that has been occurring in the main branch since release v2.5. I will keep looking but at this stage I haven't found a fix for it.

Interesting. I'll create a separate ticket for point 2 then

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]
This has come a long way since I had a chance to look at it @aldebaran30701 @Delphinus8821.

The only issue I have found with the auto save feature is that because it regenerates the UI every time, it often looses the text area I had in focus so I have to click it twice. Seleting/Deselecting layers also requires a few goes. I would suggest looking at using an "Update" button instead.

Honestly, I started to get frustrated quickly when I had to click on the same location twice or 3 times.

**Pull request merged**